### PR TITLE
[Consensus Observer] Improve peer selection and refresh logic.

### PR DESCRIPTION
**Pull request by JoshLind**
## Description
This PR improves the peer selection and refresh logic for consensus observer. Specifically, it offers these commits:
1. Move the subscription code to a separate file. This should be a no-op.
2. Improve peer selection for CO subscriptions. Here, we sort peers by validator distance and ping latency. If a previous subscription was terminated prematurely, that peer is excluded from sorting (to prevent us from selecting the same misbehaving peers). Several simple unit tests were added. More are required.
3. Add a new peer optimality check. Here, we periodically check if new (more optimal) peers would be better to subscribe to. If so, we terminate the current subscription and switch peers. This is useful to recover from failures, e.g., a validator fails, so its VFN subscribes to another peer, and then when the validator comes back up the VFN can switch back.

Note: once this lands, we have basic PFN support. For example, see this forge run: https://github.com/aptos-labs/aptos-core/actions/runs/9538849499

## Testing Plan
New and existing tests.

### CODE REVIEW

**Review by brianolson — APPROVED** [review]

File: consensus/src/consensus_observer/observer.rs [hunk 1]
```diff
@@ -285,14 +163,22 @@ impl ConsensusObserver {
...
-        if self.active_observer_subscription.is_some() {
+        // Get the peer ID of the currently active subscription
+        let active_subscription_peer = self
+            .active_observer_subscription
+            .as_ref()
+            .map(|subscription| subscription.get_peer_network_id());
+
+        // If we have an active subscription, verify that the subscription
+        // is still healthy. If not, the subscription will be terminated.
+        if active_subscription_peer.is_some() {
```
Comment (zekun000) [inline consensus/src/consensus_observer/observer.rs:174]: nit: this and the one below can be a match expression?

File: consensus/src/consensus_observer/observer.rs [hunk 2]
```diff
@@ -805,6 +725,29 @@ impl ConsensusObserver {
...
     }
 
+    /// Produces a list of sorted peers to service our subscription request. Peers
+    /// are prioritized by validator distance and latency. If `previous_subscription_peer`
+    /// is provided, it will be excluded from the selection process.
+    fn sort_peers_for_subscription(
+        &mut self,
+        previous_subscription_peer: Option<PeerNetworkId>,
+    ) -> Option<Vec<PeerNetworkId>> {
+        if let Some(mut peers_and_metadata) = self.get_connected_peers_and_metadata() {
```
Comment (zekun000) [inline consensus/src/consensus_observer/observer.rs:735]: I remember @ibalajiarun figured that cloning the HashMap is quite expensive (>10ms), we probably should use reference if possible

**Review by zekun000 — APPROVED** [review]

File: consensus/src/consensus_observer/observer.rs (hunk 1 above)
Comment (JoshLind) [reply consensus/src/consensus_observer/observer.rs:174]: Aah, I want them both to run, i.e., it's possible that the first statement terminates the subscription, and the second then creates a new one 😄 FWIW, I've refactored this in an upcoming PR, so hopefully it's a little cleaner. Will defer to that.

File: consensus/src/consensus_observer/observer.rs (hunk 2 above)
Comment (JoshLind) [reply consensus/src/consensus_observer/observer.rs:735]: Yeah, makes sense. Thankfully we're only doing this (at most) every 5 seconds, so we should be good. But, I'll see if I can clean it up in the future. 😄

**Pull request merged**